### PR TITLE
Port final prototype section of Growing guide

### PR DESCRIPTION
**Pull request by marcusramberg**

### CODE REVIEW

File: docs/Growing.md [hunk 1]
```diff
@@ -421,6 +421,113 @@ return ctx.redirectTo('goodbye');
...
+  // Store username in session
+  const session = await ctx.session();
+  session.user = ctx.stash.user;
+
+  // Store a friendly message for the next page in flash
+  const flash = await ctx.flash();
+  flash.message = 'Thanks for logging in.';
+
+  // Redirect to protected page with a 302 response
+  ctx.redirectTo('protected');
```
Comment (kraih) [inline docs/Growing.md:458]: This should `await`.

File: docs/Growing.md [hunk 2]
```diff
@@ -421,6 +421,113 @@ return ctx.redirectTo('goodbye');
...
+// Main login action
+app.any('/', async ctx => {
+  // Query parameters
+  const params = await ctx.params();
+  ctx.stash.user = params.get('user')
+  const pass = params.get('pass')
+
+  // Check password and render "index.html.ep" if necessary
+  if (ctx.models.users.check(ctx.stash.user, pass) === false) { 
+    return ctx.render({inline: indexTemplate, inlineLayout: defaultLayout});
```
Comment (kraih) [inline docs/Growing.md:446]: I think `return await` would probably be better for a doc example.

File: docs/Growing.md [hunk 3]
```diff
@@ -421,6 +421,113 @@ return ctx.redirectTo('goodbye');
...
+  return ctx.render({inline: protectedTemplate, inlineLayout: defaultLayout});
+});
+
+// Logout action
+app.get ('/logout', async ctx => {
+  // Expire and in turn clear session automatically
+  const session = await ctx.session();
+  session.expires = 1;
+  // Redirect to main page with a 302 response
+  ctx.redirectTo('/');
```
Comment (kraih) [inline docs/Growing.md:481]: `await` missing.

File: docs/Growing.md [hunk 4]
```diff
@@ -421,6 +421,113 @@ return ctx.redirectTo('goodbye');
...
+  <input name="user">
+  <br>Password:<br>
+  <input type="password" name="pass">
+  <br>
+<%= ctx.tag('button', { type: 'Submit'}, 'Login') %>
+</form>
+`;
+
+const protectedTemplate = `
+<% if (ctx.stash.flash.message != null) { %>
```
Comment (kraih) [inline docs/Growing.md:501]: You can now use JavaScript lines too, like `% if ... {`.

File: docs/Growing.md [hunk 5]
```diff
@@ -421,6 +421,113 @@ return ctx.redirectTo('goodbye');
...
+  <b><%= ctx.stash.flash.message %></b><br>
+<% } %>
+  Welcome <%= ctx.stash.session.user %>.<br>
+  <a href="/logout">Logout</a>
```
Comment (kraih) [inline docs/Growing.md:505]: Indentation is off.

File: docs/Growing.md [hunk 6]
```diff
@@ -421,6 +421,113 @@ return ctx.redirectTo('goodbye');
...
+`;
+
+const defaultLayout = `
+  <!DOCTYPE html>
+  <html>
+    <head><title>Login Manager</title></head>
+    <body><%== view.content %></body>
+  </html>
```
Comment (kraih) [inline docs/Growing.md:513]: Indentation.

File: docs/Growing.md [hunk 7]
```diff
@@ -421,6 +421,113 @@ return ctx.redirectTo('goodbye');
...
+```
+  myapp
+  |- myapp.js
+  |- models
+  |  +- users.js
+  +- tests
+     +- login.js
+```
+
+Our templates are using quite a few features of the renderer, the `Rendering` guide explains them all in
```
Comment (kraih) [inline docs/Growing.md:528]: You can link to the rendering guide.

**Comment by marcusramberg** [comment]
Addressed all the review comments now.

File: docs/Growing.md [hunk 8]
```diff
@@ -415,12 +415,119 @@ with `ctx.redirectTo()`, you can use the flash, accessible through `ctx.flash`.
...
 
+## Final prototype
+
+A final `myapp.js` prototype passing all of the tests above could look like this.
+
+```js
+import mojo from '@mojojs/core';
+import Users from './models/users.js';
+
+// Set cookie secret
```
Comment (kraih) [inline docs/Growing.md:432]: The original says `# Make signed cookies tamper resistant`. Which sounds much better.

Comment (kraih) [reply docs/Growing.md:432]: Note that cookies are encrypted here though.

File: docs/Growing.md [hunk 9]
```diff
@@ -415,12 +415,119 @@ with `ctx.redirectTo()`, you can use the flash, accessible through `ctx.flash`.
...
+export const app = mojo({secrets: ['Mojolicious rocks']});
+
+app.models.users = new Users();
+
+// Main login action
+app.any('/', async ctx => {
+  // Query parameters
```
Comment (kraih) [inline docs/Growing.md:439]: `Query or POST parameters`.

File: docs/Growing.md [hunk 10]
```diff
@@ -415,12 +415,119 @@ with `ctx.redirectTo()`, you can use the flash, accessible through `ctx.flash`.
...
+  const params = await ctx.params();
+  ctx.stash.user = params.get('user')
+  const pass = params.get('pass')
+
+  // Check password and render "index.html.ep" if necessary
```
Comment (kraih) [inline docs/Growing.md:444]: `index.html.mt`.

File: docs/Growing.md (hunk 8 above)
Comment (marcusramberg) [reply docs/Growing.md:432]: I know, that's why I changed it from tamper resistant, which isn't really true anymore once they are encrypted.

File: docs/Growing.md (hunk 10 above)
Comment (kraih) [reply docs/Growing.md:444]: No wait, it's actually inline templates without name.

Comment (marcusramberg) [reply docs/Growing.md:444]: 'the inline index template' sound ok to you?

File: docs/Growing.md [hunk 11]
```diff
@@ -415,12 +415,119 @@ with `ctx.redirectTo()`, you can use the flash, accessible through `ctx.flash`.
...
+
+// Make sure user is logged in for actions in this action
+const protectedArea = app.under('/protected').to(async ctx => {
+  ctx.stash.session = await ctx.session();
+  if (ctx.stash.session.user !== undefined) return;
+  await ctx.render({text: 'Permission denied'});
+  return false;
+});
+
+// A protected page auto rendering "protected.html.ep"
```
Comment (kraih) [inline docs/Growing.md:469]: Also an inline template.

File: docs/Growing.md [hunk 12]
```diff
@@ -415,12 +415,119 @@ with `ctx.redirectTo()`, you can use the flash, accessible through `ctx.flash`.
...
+
+  // Redirect to protected page with a 302 response
+  await ctx.redirectTo('protected');
+}); 
+
+// Make sure user is logged in for actions in this action
+const protectedArea = app.under('/protected').to(async ctx => {
+  ctx.stash.session = await ctx.session();
+  if (ctx.stash.session.user !== undefined) return;
+  await ctx.render({text: 'Permission denied'});
```
Comment (kraih) [inline docs/Growing.md:465]: Isn't this a redirect in the original?

File: docs/Growing.md [hunk 13]
```diff
@@ -415,12 +415,119 @@ with `ctx.redirectTo()`, you can use the flash, accessible through `ctx.flash`.
...
+  return false;
+});
+
+// A protected page auto rendering "protected.html.ep"
+protectedArea.get('/').to(async ctx => {
+  ctx.stash.flash = await ctx.flash();
```
Comment (kraih) [inline docs/Growing.md:471]: This is rather clunky. Templates also have access to `ctx.flash()`.

File: docs/Growing.md [hunk 14]
```diff
@@ -415,12 +415,119 @@ with `ctx.redirectTo()`, you can use the flash, accessible through `ctx.flash`.
...
+  <input name="user">
+  <br>Password:<br>
+  <input type="password" name="pass">
+  <br>
+<%= ctx.tag('button', { type: 'Submit'}, 'Login') %>
+</form>
+`;
+
+const protectedTemplate = `
+% if (ctx.stash.flash.message != null) {
```
Comment (kraih) [inline docs/Growing.md:501]: Writing out `ctx.stash.flash.message` kinda defeats the purpose of assigning anything to the stash in the first place.

File: docs/Growing.md [hunk 15]
```diff
@@ -415,12 +415,119 @@ with `ctx.redirectTo()`, you can use the flash, accessible through `ctx.flash`.
...
+<form method="post">
+<% if (ctx.stash.user !== null) { %>
+  <b>Wrong name or password, please try again.</b><br>
+<% } %>
+  User:<br>
+  <input name="user">
+  <br>Password:<br>
+  <input type="password" name="pass">
+  <br>
+<%= ctx.tag('button', { type: 'Submit'}, 'Login') %>
```
Comment (kraih) [inline docs/Growing.md:496]: The helper use doesn't really add anything to the example.

File: docs/Growing.md (hunk 14 above)
Comment (kraih) [reply docs/Growing.md:501]: Simply adding a `% const flash = await ctx.flash();` to the top would be cleaner.

File: docs/Growing.md [hunk 16]
```diff
@@ -415,12 +415,119 @@ with `ctx.redirectTo()`, you can use the flash, accessible through `ctx.flash`.
...
+```
+  myapp
+  |- myapp.js
+  |- models
+  |  +- users.js
+  +- tests
+     +- login.js
+```
+
+Our templates are using quite a few features of the renderer, the [Rendering](Rendering.md)` guide explains them all in
```
Comment (kraih) [inline docs/Growing.md:528]: That "`" doesn't belong.

File: docs/Growing.md [hunk 17]
```diff
@@ -415,12 +415,119 @@ with `ctx.redirectTo()`, you can use the flash, accessible through `ctx.flash`.
...
+
+And the directory structure should be looking like this now.
+
+```
+  myapp
+  |- myapp.js
+  |- models
+  |  +- users.js
+  +- tests
+     +- login.js
```
Comment (kraih) [inline docs/Growing.md:525]: Indentation.

File: docs/Growing.md (hunk 8 above)
Comment (kraih) [reply docs/Growing.md:432]: `Make encrypted cookies more secure with custom secrets` maybe?

File: docs/Growing.md (hunk 10 above)
Comment (kraih) [reply docs/Growing.md:444]: "index inline template"?

File: docs/Growing.md (hunk 14 above)
Comment (kraih) [reply docs/Growing.md:501]: Templates are always `async`, so you can use `await` freely.

File: docs/Growing.md (hunk 9 above)
Comment (marcusramberg) [reply docs/Growing.md]: Fixed here and in previous use in the Growing guide.

File: docs/Growing.md [hunk 18]
```diff
@@ -415,12 +415,119 @@ with `ctx.redirectTo()`, you can use the flash, accessible through `ctx.flash`.
...
+// Set custom cookie secret to ensure encryption is more secure
+export const app = mojo({secrets: ['Mojolicious rocks']});
+
+app.models.users = new Users();
+
+// Main login action
+app.any('/', async ctx => {
+  // Query or POST parameters
+  const params = await ctx.params();
+  ctx.stash.user = params.get('user')
```
Comment (kraih) [inline docs/Growing.md:441]: Missing semicolons.

File: docs/Growing.md [hunk 19]
```diff
@@ -415,12 +415,119 @@ with `ctx.redirectTo()`, you can use the flash, accessible through `ctx.flash`.
...
+  const flash = await ctx.flash();
+  flash.message = 'Thanks for logging in.';
+
+  // Redirect to protected page with a 302 response
+  await ctx.redirectTo('protected');
+}); 
+
+// Make sure user is logged in for actions in this action
+const protectedArea = app.under('/protected').to(async ctx => {
+  ctx.stash.session = await ctx.session();
```
Comment (kraih) [inline docs/Growing.md:463]: Why are you storing the session in the stash?

File: docs/Growing.md [hunk 20]
```diff
@@ -415,12 +415,119 @@ with `ctx.redirectTo()`, you can use the flash, accessible through `ctx.flash`.
...
+  if (ctx.stash.session.user !== undefined) return;
+  await ctx.redirectTo('/');
+  return false;
+});
+
+// A protected page auto rendering the protected inline template"
+protectedArea.get('/').to(async ctx => {
+  return await ctx.render({inline: protectedTemplate, inlineLayout: defaultLayout});
```
Comment (kraih) [inline docs/Growing.md:471]: The `return` here does nothing.

**Pull request merged**